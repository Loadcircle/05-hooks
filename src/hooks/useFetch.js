import { useEffect, useRef, useState } from "react"

export const useFetch = (url, initialState = {data: null, loading: true, error: null}) => {

    const isMounted = useRef(true)

    const [state, setState] = useState(initialState);

    useEffect(()=>{

        return()=>{
            isMounted.current = false;
        }

    }, [])

    useEffect(() => {
        setState({...initialState, loading: true})
        fetch(url)
            .then(resp=>resp.json())
            .then(data=>{
                if(isMounted.current){
                    setState({
                        loading: false,
                        error: null,
                        data,
                    });
                }
            })
    }, [url])

    return state;
    
}
