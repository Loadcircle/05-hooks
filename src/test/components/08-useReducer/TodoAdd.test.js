import React from 'react';
import { useForm } from '../../../hooks/useForm';
const { shallow } = require("enzyme");
const { TodoAdd } = require("../../../components/08-useReducer/TodoAdd");

const handleAddTodo = jest.fn();
// const handleInputChange = jest.fn();
// const reset = jest.fn();
// jest.mock('../../../hooks/useForm');

// useForm.mockReturnValue({
//     description: '',
//     handleInputChange: handleInputChange,
//     reset: reset,
// });

describe('Pruebas en el todoadd', ()=>{

    const wrapper = shallow(<TodoAdd
        handleAddTodo={handleAddTodo}
    />);

    test('Debe hacer match el snapshot', ()=>{

        expect(wrapper).toMatchSnapshot();

    });

    test('No debe de llamar handleAddTodo', ()=>{

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault: ()=>{}});

        expect(handleAddTodo).toHaveBeenCalledTimes(0);

    });
    
    test('Debe de llamar la funcion handleAddTodo y el reset', ()=>{
        
        const value = 'Aprender React';
        wrapper.find('input').simulate('change', {
            target: {
                value: value,
                name: 'description',
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault: ()=>{}});
        
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));

        //le decimos que sea llamado con un objeto con estos valores y cualquier id siempre y cuando sea un numero
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number), 
            desc: value,
            done: false,            
        });        

        expect(wrapper.find('input').prop('value')).toBe('');

    });
    
    
    
});