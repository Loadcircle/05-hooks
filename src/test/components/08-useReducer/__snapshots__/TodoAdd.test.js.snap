// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el todoadd Debe hacer match el snapshot 1`] = `
<Fragment>
  <h4>
    Agregar TODO'S
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Tarea"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
